libyul/optimiser/UnusedStoreEliminator.cpp

```diff
@@ -352,30 +352,19 @@ bool UnusedStoreEliminator::knownUnrelated(
 		if (_op2.length && knowledge.knownToBeZero(*_op2.length))
 			return true;
 
-		// 1.start + 1.length <= 2.start
-		if (_op1.length && _op2.start && _op1.start)
-		{
-			optional<u256> length1 = knowledge.valueIfKnownConstant(*_op1.length);
-			optional<u256> diff = knowledge.differenceIfKnownConstant(*_op2.start, *_op1.start);
-			// diff = 2.start - 1.start
-			if (length1 && diff)
-				if (
-					*length1 <= *diff &&
-					*length1 <= largestPositive &&
-					*diff <= largestPositive
-				)
-					return true;
-		}
-		// 2.start + 2.length <= 1.start
-		if (_op2.length && _op1.start && _op2.start)
-		{
-			optional<u256> length2 = knowledge.valueIfKnownConstant(*_op2.length);
-			optional<u256> diff = knowledge.differenceIfKnownConstant(*_op1.start, *_op2.start);
-			// diff = 1.start - 2.start
-			if (length2 && diff)
-				if (*length2 <= *diff && *length2 <= largestPositive && *diff <= largestPositive)
-					return true;
-		}
+		// We use diff >= _op1.length && diff <= 2**256 - _op2.length
```

Comment:
This is so annoying. This is not enough to remove the mstore in
```
mstore(0, 0)
let length := calldataload(0)
calldatacopy(0x20, 0x20, length)
return(0x20, length)
```

Reply:
Just to check that I understand this thing correctly (which I might not yet)... what currently happens is that both the ``mstore`` and the ``calldatacopy`` are ``Undecided`` when we hit the ``return`` and then the ``return`` does not know that it won't read from ``0`` because of the potential overflow?

Couldn't we when hitting the ``return`` walk the previous memory writes in backwards order and check for ``knownCovered`` writes and if we find one stop?

I.e. ``return(0x20, length)`` would first look at ``calldatacopy(0x20, 0x20, length)``, realize that its entire read is covered by that write and be satisfied and not even consider earlier writes, i.e. not even consider the ``mstore(0,0)``...
Then the ``mstore(0,0)`` would remain ``Undecided`` and be removed, right?

Reply:
I mean: the reason that we can remove ``mstore(0,0)`` is *not* that the return won't read from offset ``0``, but that it will only read what ``calldatacopy`` wrote, isn't it?

Reply:
The order of writes is not preserved in ``m_stores``, though, is it? Hm...

Reply:
The reason is that `knownUnrelated` does not return `true` for the `mstore(0, 0)` and `return(0x20, length)`.

Reply:
We can remove `mstore(0, 0)` because nothing reads from that location until the call terminates.

Reply:
If you had ``mstore(0x20, 0)`` instead, then ``return(0x20, length)`` would *not* be unrelated, but the ``mstore`` could *still* be removed.

Reply:
I could actually imagine that that reason is the more useful one in general, since we'll probably rarely just read random stuff that wasn't written specifically... so in general, I think it will more often be the case that something only reads stuff covered by earlier writes rather than that the locations are known to be disjoint...

In general, though, we would have to track this across multiple writes - it's just a coincidence that it's a single write that covers the entire read in this example... i.e. in

```
mstore(y, 0x42)
mstore(x, 0x20)
mstore(add(x, 0x20), 0x20)
return(x, 0x40)
```

I'd ideally want to recognize that the last two ``mstore``s together cover everything that's read by the ``return`` and be able to remove the first ``mstore`` without any knowledge about ``x`` or ``y``... but yeah, just ideally, this may be infeasible to do...

Reply:
Well, actually, probably not infeasible... but at least tricky :-).

Reply:
Note from call: maybe it's worth differentiating between ``diff`` being positive or negative. So with an ``s256 diff``, it may be
```
if (diff >= 0 && diff >= length1)
    return true;
if (diff < 0 && -diff >= length2)
    return true;
```

But at least I haven't convinced myself yet that there's no other weird overflow cases to consider.